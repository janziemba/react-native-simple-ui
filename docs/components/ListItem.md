---
id: listitem
title: ListItem
---
## Props
- `centerElement` (_node_, default null)
- `centerElementSize` (_number_, default 1.5)
- `color` (_string_, default white) - one of theme colors, [base theme colors](../Theme.md#colors) by default
- `divider` (_boolean_, default true)
- `leftElement` (_node_, default null)
- `leftElementSize` (_number_, default 1)
- `loading` (_boolean_, default true)
- `rightElement` (_node_, default null)
- `rightElementSize` (_number_, default 1)

<img src="../assets/ListItem/1.png" alt="ListItem" width="250" style="display: inline-block;">
