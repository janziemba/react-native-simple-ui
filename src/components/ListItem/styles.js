// @flow

import { StyleSheet } from 'react-native';

import merge from 'lodash/merge';

import type { ObjectOfStringsType, ObjectOfStyleSheetsType, ThemeShapeType } from '../../types';
import { COLORS } from './constants';

export type StylesType = {
    base: {
        centerElement: StyleSheet.Styles,
        innerContainer: StyleSheet.Styles,
        leftElement: StyleSheet.Styles,
        rightElement: StyleSheet.Styles,
    },
    colors: ObjectOfStyleSheetsType,
};

export default ({ colors }: ThemeShapeType): StylesType => {
    const allColors: ObjectOfStringsType = merge({}, COLORS, colors.palette);
    const colorOptions: ObjectOfStyleSheetsType = {};

    Object.keys(allColors).forEach((colorName: string) => {
        const color: string = colors.palette[colorName] || colorName;

        colorOptions[colorName] = {
            outerContainer: {
                backgroundColor: color,
            },
        };
    });

    return {
        base: {
            centerElement: {
                flexDirection: 'row',
                justifyContent: 'center',
            },
            innerContainer: {
                base: {
                    alignItems: 'center',
                    height: 56,
                    justifyContent: 'center',
                },
            },
            leftElement: {
                flexDirection: 'row',
            },
            rightElement: {
                flexDirection: 'row',
                justifyContent: 'flex-end',
            },
        },
        colors: colorOptions,
    };
};
