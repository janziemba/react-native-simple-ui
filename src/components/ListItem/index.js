// @flow

import React, { PureComponent } from 'react';
import type { Node } from 'react';
import { ActivityIndicator, StyleSheet, TouchableOpacity, View } from 'react-native';

import omit from 'lodash/omit';
import type { Props as TouchableOpacityProps } from 'react-native/Libraries/Components/Touchable/TouchableOpacity';

import withTheme from '../../themes/withTheme';
import hasStyleChanged from '../../utils/hasStyleChanged';
import Divider from '../Divider';
import Column from '../Grid/Column';
import Grid from '../Grid/Grid';
import Padding from '../Padding';
import styles from './styles';
import type { StylesType } from './styles';

type Props = TouchableOpacityProps & {
    centerElement?: Node,
    centerElementSize?: number, // eslint-disable-line react/no-unused-prop-types
    color?: string, // eslint-disable-line react/no-unused-prop-types
    divider?: boolean, // eslint-disable-line react/no-unused-prop-types
    leftElement?: Node,
    leftElementSize?: number, // eslint-disable-line react/no-unused-prop-types
    loading?: boolean,
    rightElement?: Node,
    rightElementSize?: number, // eslint-disable-line react/no-unused-prop-types
    styles: StylesType, // eslint-disable-line react/no-unused-prop-types
};

type MergedStylesType = {
    centerElement: StyleSheet.Styles,
    innerContainer: Array<StyleSheet.Styles>,
    leftElement: StyleSheet.Styles,
    outerContainer: Array<StyleSheet.Styles>,
    rightElement: StyleSheet.Styles,
};

type State = {
    styles: MergedStylesType,
};

const getStyles = (props: Props): MergedStylesType => {
    const { color, styles } = props;

    return {
        ...styles.base,
        outerContainer: [
            styles.base.outerContainer,
            styles.colors[color || 'white'].outerContainer,
        ],
    };
};

class ListItem extends PureComponent<Props, State> {
    static defaultProps = {
        centerElement: null,
        centerElementSize: 1.5,
        color: 'white',
        divider: true,
        leftElement: null,
        leftElementSize: 1,
        loading: false,
        rightElement: null,
        rightElementSize: 1,
    };
    state = {
        styles: getStyles(this.props),
    };
    constructor(props: Props): void {
        super(props);

        (this: any).renderLeftElement = this.renderLeftElement.bind(this);
        (this: any).renderCenterElement = this.renderCenterElement.bind(this);
        (this: any).renderRightElement = this.renderRightElement.bind(this);
    }
    componentWillReceiveProps(nextProps: Props): void {
        const propsOnWhichDependsTheStyle: Array<string> = ['color'];

        if (hasStyleChanged(propsOnWhichDependsTheStyle, nextProps, this.props)) {
            this.setState({ styles: getStyles(nextProps) });
        }
    }
    renderLeftElement(): Node {
        const { leftElement, loading } = this.props;
        const { styles } = this.state;

        if (!leftElement || loading) {
            return null;
        }

        return (
            <Padding
                size="small"
                style={styles.leftElement}
            >
                {leftElement}
            </Padding>
        );
    }
    renderCenterElement(): Node {
        const { centerElement, loading } = this.props;
        const { styles } = this.state;

        if (loading) {
            return (
                <ActivityIndicator />
            );
        }

        if (!centerElement) {
            return null;
        }

        return (
            <Padding
                size="small"
                style={styles.centerElement}
            >
                {centerElement}
            </Padding>
        );
    }
    renderRightElement(): Node {
        const { loading, rightElement } = this.props;
        const { styles } = this.state;

        if (!rightElement || loading) {
            return null;
        }

        return (
            <Padding
                size="small"
                style={styles.rightElement}
            >
                {rightElement}
            </Padding>
        );
    }
    renderDivider(): Node {
        const { divider } = this.props;

        if (!divider) {
            return null;
        }

        return <Divider />;
    }
    render(): Node {
        const { centerElementSize, leftElementSize, onPress, rightElementSize } = this.props;
        const { styles } = this.state;

        return (
            <View style={styles.outerContainer}>
                <TouchableOpacity
                    disabled={!onPress}
                    onPress={onPress}
                    {...omit(this.props, ['styles'])}
                >
                    <Grid style={styles.innerContainer}>
                        <Column size={leftElementSize}>
                            {this.renderLeftElement()}
                        </Column>
                        <Column size={centerElementSize}>
                            {this.renderCenterElement()}
                        </Column>
                        <Column size={rightElementSize}>
                            {this.renderRightElement()}
                        </Column>
                    </Grid>
                </TouchableOpacity>
                {this.renderDivider()}
            </View>
        );
    }
}

ListItem = withTheme(styles, 'ListItem')(ListItem);

export default ListItem;
