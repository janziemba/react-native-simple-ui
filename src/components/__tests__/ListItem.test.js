import React from 'react';
import renderer from 'react-test-renderer';

import ListItem from '../ListItem';

it('renders correctly', () => {
    const tree = renderer.create(<ListItem />).toJSON();
    expect(tree).toMatchSnapshot();
});
