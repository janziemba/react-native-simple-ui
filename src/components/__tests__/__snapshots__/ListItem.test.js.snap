// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      undefined,
      Object {
        "backgroundColor": "white",
      },
    ]
  }
>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexShrink": 1,
            "height": 56,
            "justifyContent": "center",
          },
          Object {
            "flexDirection": "row",
          },
        ]
      }
      styles={
        Object {
          "base": Object {
            "alignItems": "center",
            "flexShrink": 1,
            "height": 56,
            "justifyContent": "center",
          },
          "directions": Object {
            "column": Object {
              "flexDirection": "column",
            },
            "row": Object {
              "flexDirection": "row",
            },
          },
        }
      }
      theme={
        Object {
          "colors": Object {
            "layout": Object {
              "backgroundColor": "#ffffff",
            },
            "palette": Object {
              "danger": "#ff4136",
              "dark": "#484848",
              "disabled": "#999999",
              "info": "#75caeb",
              "light": "#f6f6f6",
              "primary": "#158cba",
              "success": "#28b62c",
              "warning": "#ff851b",
            },
            "text": Object {
              "disabled": "#999999",
              "placeholder": "#e2e4e6",
              "primary": "#3d3d3d",
              "secondary": "#9e9e9e",
            },
          },
          "components": Object {},
          "radiuses": Object {
            "gigantic": 16,
            "huge": 12,
            "large": 8,
            "medium": 6,
            "small": 4,
            "tiny": 2,
          },
          "spacing": Object {
            "gigantic": 24,
            "huge": 20,
            "large": 16,
            "medium": 12,
            "small": 8,
            "tiny": 4,
          },
          "typography": Object {
            "fontFamily": "System",
            "fontSizes": Object {
              "large": 16,
              "medium": 14,
              "small": 12,
            },
            "fontWeights": Object {
              "bold": "900",
              "medium": "800",
              "normal": "600",
            },
            "iconSet": "MaterialIcons",
            "sizes": Object {
              "h1": Object {
                "fontSize": 31,
                "fontWeight": "900",
              },
              "h2": Object {
                "fontSize": 25,
                "fontWeight": "900",
              },
              "h3": Object {
                "fontSize": 19,
                "fontWeight": "800",
              },
              "h4": Object {
                "fontSize": 17,
                "fontWeight": "800",
              },
              "h5": Object {
                "fontSize": 16,
                "fontWeight": "800",
              },
              "h6": Object {
                "fontSize": 15,
                "fontWeight": "700",
              },
              "normal": Object {
                "fontSize": 14,
                "fontWeight": "700",
              },
              "small": Object {
                "fontSize": 12,
              },
            },
          },
        }
      }
    >
      <View
        size={1}
        style={
          Array [
            Object {
              "flexDirection": "column",
            },
            Object {
              "flex": 1,
            },
          ]
        }
        styles={
          Object {
            "base": Object {
              "flexDirection": "column",
            },
          }
        }
        theme={
          Object {
            "colors": Object {
              "layout": Object {
                "backgroundColor": "#ffffff",
              },
              "palette": Object {
                "danger": "#ff4136",
                "dark": "#484848",
                "disabled": "#999999",
                "info": "#75caeb",
                "light": "#f6f6f6",
                "primary": "#158cba",
                "success": "#28b62c",
                "warning": "#ff851b",
              },
              "text": Object {
                "disabled": "#999999",
                "placeholder": "#e2e4e6",
                "primary": "#3d3d3d",
                "secondary": "#9e9e9e",
              },
            },
            "components": Object {},
            "radiuses": Object {
              "gigantic": 16,
              "huge": 12,
              "large": 8,
              "medium": 6,
              "small": 4,
              "tiny": 2,
            },
            "spacing": Object {
              "gigantic": 24,
              "huge": 20,
              "large": 16,
              "medium": 12,
              "small": 8,
              "tiny": 4,
            },
            "typography": Object {
              "fontFamily": "System",
              "fontSizes": Object {
                "large": 16,
                "medium": 14,
                "small": 12,
              },
              "fontWeights": Object {
                "bold": "900",
                "medium": "800",
                "normal": "600",
              },
              "iconSet": "MaterialIcons",
              "sizes": Object {
                "h1": Object {
                  "fontSize": 31,
                  "fontWeight": "900",
                },
                "h2": Object {
                  "fontSize": 25,
                  "fontWeight": "900",
                },
                "h3": Object {
                  "fontSize": 19,
                  "fontWeight": "800",
                },
                "h4": Object {
                  "fontSize": 17,
                  "fontWeight": "800",
                },
                "h5": Object {
                  "fontSize": 16,
                  "fontWeight": "800",
                },
                "h6": Object {
                  "fontSize": 15,
                  "fontWeight": "700",
                },
                "normal": Object {
                  "fontSize": 14,
                  "fontWeight": "700",
                },
                "small": Object {
                  "fontSize": 12,
                },
              },
            },
          }
        }
      />
      <View
        size={1.5}
        style={
          Array [
            Object {
              "flexDirection": "column",
            },
            Object {
              "flex": 1.5,
            },
          ]
        }
        styles={
          Object {
            "base": Object {
              "flexDirection": "column",
            },
          }
        }
        theme={
          Object {
            "colors": Object {
              "layout": Object {
                "backgroundColor": "#ffffff",
              },
              "palette": Object {
                "danger": "#ff4136",
                "dark": "#484848",
                "disabled": "#999999",
                "info": "#75caeb",
                "light": "#f6f6f6",
                "primary": "#158cba",
                "success": "#28b62c",
                "warning": "#ff851b",
              },
              "text": Object {
                "disabled": "#999999",
                "placeholder": "#e2e4e6",
                "primary": "#3d3d3d",
                "secondary": "#9e9e9e",
              },
            },
            "components": Object {},
            "radiuses": Object {
              "gigantic": 16,
              "huge": 12,
              "large": 8,
              "medium": 6,
              "small": 4,
              "tiny": 2,
            },
            "spacing": Object {
              "gigantic": 24,
              "huge": 20,
              "large": 16,
              "medium": 12,
              "small": 8,
              "tiny": 4,
            },
            "typography": Object {
              "fontFamily": "System",
              "fontSizes": Object {
                "large": 16,
                "medium": 14,
                "small": 12,
              },
              "fontWeights": Object {
                "bold": "900",
                "medium": "800",
                "normal": "600",
              },
              "iconSet": "MaterialIcons",
              "sizes": Object {
                "h1": Object {
                  "fontSize": 31,
                  "fontWeight": "900",
                },
                "h2": Object {
                  "fontSize": 25,
                  "fontWeight": "900",
                },
                "h3": Object {
                  "fontSize": 19,
                  "fontWeight": "800",
                },
                "h4": Object {
                  "fontSize": 17,
                  "fontWeight": "800",
                },
                "h5": Object {
                  "fontSize": 16,
                  "fontWeight": "800",
                },
                "h6": Object {
                  "fontSize": 15,
                  "fontWeight": "700",
                },
                "normal": Object {
                  "fontSize": 14,
                  "fontWeight": "700",
                },
                "small": Object {
                  "fontSize": 12,
                },
              },
            },
          }
        }
      />
      <View
        size={1}
        style={
          Array [
            Object {
              "flexDirection": "column",
            },
            Object {
              "flex": 1,
            },
          ]
        }
        styles={
          Object {
            "base": Object {
              "flexDirection": "column",
            },
          }
        }
        theme={
          Object {
            "colors": Object {
              "layout": Object {
                "backgroundColor": "#ffffff",
              },
              "palette": Object {
                "danger": "#ff4136",
                "dark": "#484848",
                "disabled": "#999999",
                "info": "#75caeb",
                "light": "#f6f6f6",
                "primary": "#158cba",
                "success": "#28b62c",
                "warning": "#ff851b",
              },
              "text": Object {
                "disabled": "#999999",
                "placeholder": "#e2e4e6",
                "primary": "#3d3d3d",
                "secondary": "#9e9e9e",
              },
            },
            "components": Object {},
            "radiuses": Object {
              "gigantic": 16,
              "huge": 12,
              "large": 8,
              "medium": 6,
              "small": 4,
              "tiny": 2,
            },
            "spacing": Object {
              "gigantic": 24,
              "huge": 20,
              "large": 16,
              "medium": 12,
              "small": 8,
              "tiny": 4,
            },
            "typography": Object {
              "fontFamily": "System",
              "fontSizes": Object {
                "large": 16,
                "medium": 14,
                "small": 12,
              },
              "fontWeights": Object {
                "bold": "900",
                "medium": "800",
                "normal": "600",
              },
              "iconSet": "MaterialIcons",
              "sizes": Object {
                "h1": Object {
                  "fontSize": 31,
                  "fontWeight": "900",
                },
                "h2": Object {
                  "fontSize": 25,
                  "fontWeight": "900",
                },
                "h3": Object {
                  "fontSize": 19,
                  "fontWeight": "800",
                },
                "h4": Object {
                  "fontSize": 17,
                  "fontWeight": "800",
                },
                "h5": Object {
                  "fontSize": 16,
                  "fontWeight": "800",
                },
                "h6": Object {
                  "fontSize": 15,
                  "fontWeight": "700",
                },
                "normal": Object {
                  "fontSize": 14,
                  "fontWeight": "700",
                },
                "small": Object {
                  "fontSize": 12,
                },
              },
            },
          }
        }
      />
    </View>
  </View>
  <View
    color={null}
    style={
      Array [
        Object {
          "backgroundColor": "#484848",
          "height": 0.5,
        },
        Object {},
      ]
    }
    styles={
      Object {
        "base": Object {
          "backgroundColor": "#484848",
          "height": 0.5,
        },
        "colors": Object {
          "black": Object {
            "backgroundColor": "black",
          },
          "danger": Object {
            "backgroundColor": "#ff4136",
          },
          "dark": Object {
            "backgroundColor": "#484848",
          },
          "disabled": Object {
            "backgroundColor": "#999999",
          },
          "info": Object {
            "backgroundColor": "#75caeb",
          },
          "light": Object {
            "backgroundColor": "#f6f6f6",
          },
          "primary": Object {
            "backgroundColor": "#158cba",
          },
          "success": Object {
            "backgroundColor": "#28b62c",
          },
          "warning": Object {
            "backgroundColor": "#ff851b",
          },
          "white": Object {
            "backgroundColor": "white",
          },
        },
      }
    }
    theme={
      Object {
        "colors": Object {
          "layout": Object {
            "backgroundColor": "#ffffff",
          },
          "palette": Object {
            "danger": "#ff4136",
            "dark": "#484848",
            "disabled": "#999999",
            "info": "#75caeb",
            "light": "#f6f6f6",
            "primary": "#158cba",
            "success": "#28b62c",
            "warning": "#ff851b",
          },
          "text": Object {
            "disabled": "#999999",
            "placeholder": "#e2e4e6",
            "primary": "#3d3d3d",
            "secondary": "#9e9e9e",
          },
        },
        "components": Object {},
        "radiuses": Object {
          "gigantic": 16,
          "huge": 12,
          "large": 8,
          "medium": 6,
          "small": 4,
          "tiny": 2,
        },
        "spacing": Object {
          "gigantic": 24,
          "huge": 20,
          "large": 16,
          "medium": 12,
          "small": 8,
          "tiny": 4,
        },
        "typography": Object {
          "fontFamily": "System",
          "fontSizes": Object {
            "large": 16,
            "medium": 14,
            "small": 12,
          },
          "fontWeights": Object {
            "bold": "900",
            "medium": "800",
            "normal": "600",
          },
          "iconSet": "MaterialIcons",
          "sizes": Object {
            "h1": Object {
              "fontSize": 31,
              "fontWeight": "900",
            },
            "h2": Object {
              "fontSize": 25,
              "fontWeight": "900",
            },
            "h3": Object {
              "fontSize": 19,
              "fontWeight": "800",
            },
            "h4": Object {
              "fontSize": 17,
              "fontWeight": "800",
            },
            "h5": Object {
              "fontSize": 16,
              "fontWeight": "800",
            },
            "h6": Object {
              "fontSize": 15,
              "fontWeight": "700",
            },
            "normal": Object {
              "fontSize": 14,
              "fontWeight": "700",
            },
            "small": Object {
              "fontSize": 12,
            },
          },
        },
      }
    }
  />
</View>
`;
